v2/tools/generator/internal/codegen/pipeline/promote_root_oneof-properties.go

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+
+	"github.com/rotisserie/eris"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/Azure/azure-service-operator/v2/internal/set"
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astmodel"
+)
+
+const promoteRootOneOfProperties = "promoteRootOneOfProperties"
+
+func PromoteRootOneOfProperties() *Stage {
+	stage := NewStage(
+		promoteRootOneOfProperties,
+		"Promote selected properties on OneOf objects directly referenced by resources",
+		func(ctx context.Context, state *State) (*State, error) {
+			promoter := newRootPropertyPromoter(state.Definitions())
+			newDefs, err := promoter.modifyDefinitions()
+			if err != nil {
+				return nil, eris.Wrapf(err, "assembling OneOf types")
+			}
+
+			return state.WithOverlaidDefinitions(newDefs), nil
+		})
+
+	return stage
+}
+
+type rootPropertyPromoter struct {
+	definitions          astmodel.TypeDefinitionSet
+	promotableProperties set.Set[string]
+	rootVisitor          astmodel.TypeVisitor[astmodel.TypeDefinitionSet]
+	leafVisitor          astmodel.TypeVisitor[astmodel.PropertySet]
+}
+
+func newRootPropertyPromoter(
+	defs astmodel.TypeDefinitionSet,
+) *rootPropertyPromoter {
+	result := &rootPropertyPromoter{
+		definitions:          defs,             // All known definitions
+		promotableProperties: set.Make("Name"), // Properties we want to promote (hard coded for now)
```

Comment:
Should we consider opening an issue for tracking this hard-coded prop?

Reply:
Good idea. #4597